design/ingressroute-v1-design.md

```diff
@@ -0,0 +1,320 @@
+    - This regex string is a regular expression rule which implies the entire request header value must match the regex.
+      The rule will not match if only a subsequence of the request header value matches the regex.
+      The regex grammar used in the value field is defined [here](https://en.cppreference.com/w/cpp/regex/ecmascript).
+  - Allow delegation from fqdn to header or path
+    - Sub-delegation to additional header or path
+  - Secure Backend: The idea to proxy to a backend service over a TLS connection  (Currently an annotation on the Kubernetes service)
+  - Backend Protocol: Proxy to http/2 backend (Currently an annoation on the Kubernetes service)
+
+## Non Goals
+
+- **<u>Overall</u>**:
+  - Deprecate Contour's support for v1beta1.Ingress.
+  - Support Ingress objects outside the current cluster.
+  - IP in IP or GRE tunneling between clusters with discontinuous or overlapping IP ranges.
+- **<u>Routing:</u>**
+  - Support JWT tokens with header routing
+
+## High-Level Design
+
+At a high level, this document proposes modeling ingress configuration as a graph of documents throughout a Kubernetes API server, which when taken together form a directed acyclic graph (DAG) of the configuration for virtual hosts, request headers, and their constituent routes.
+
+The v1.IngressRoute design looks to add header based routing requiring changes to v1beta1.IngressRoute to allow for a spec to define the key/value pairs to match against. This change updates how IngressRoute is structured to allow for specifying both path and header per route match. 
+
+### Delegation
+
+The working model for delegation is DNS. As the owner of a DNS domain, for example `.com`, I *delegate* to another nameserver the responsibility for handing the subdomain `heptio.com`. Any nameserver can hold a record for `heptio.com`, but without the linkage from the parent `.com` TLD, its information is unreachable and non authoritative.
+
+Each *root* of a DAG starts at a virtual host, which describes properties such as the fully qualified name of the virtual host, TLS configuration, and possibly global access list details. The vertices of a graph do not contain virtual host information. Instead they are reachable from a root only by delegation. This permits the *owner* of an ingress root to both delegate the authority to publish a service on a portion of the route space (path or request header) inside a virtual host, and to further delegate authority to publish and delegate.
+
+In practice the linkage, or delegation, from root to vertex, is performed with a specific type of route action. You can think of it as routing traffic to another ingress route for further processing, instead of routing traffic directly to a service.
+
+## Detailed Design
+
+### Changes from v1beta1
+
+Here are some changes that are directly different from the v1beta1 design:
+
+- `services` under the `match` struct are renamed to `backends`
```

Comment:
Looks like you may need to `s/services/backends/g` - some of the sample IngressRoutes below are still using `services:`.

Reply:
👍